### ISSUE TITLE
When using cnpm/pnpm, rg uses lots of CPU

### ISSUE THREAD

**Issue by 617450941** [Microsoft/vscode#35659]
update vscode 1.17.0 CPU usage at 100%  rg.exe  bug?

**Comment by ffxsam** [comment]
Your report isn't very clear (not sure what rg.exe is), but on Ubuntu, I'm getting 100% cpu spikes just from typing.

I'm desperately looking for a way to download the version previous to this one.

**Comment by 617450941** [comment]
file location  \Microsoft VS Code\resources\app\node_modules\vscode-ripgrep\bin\rg.exe

**Comment by 617450941** [comment]
![image](https://user-images.githubusercontent.com/7607957/31259468-73ac9678-aa79-11e7-9d3a-e1b25b827409.png)

**Comment by 617450941** [comment]
![image](https://user-images.githubusercontent.com/7607957/31259498-acaa52e4-aa79-11e7-96a4-d52e5478340e.png)

**Comment by cipchk** [comment]
run as administrator the problem still exists.

**Comment by chrmarti** [comment]
Could you check what the command arguments are?

**Comment by chrmarti** [comment]
Also: Does this happen immediately after opening the window or only after using Ctrl + P to show QuickOpen?

**Comment by 617450941** [comment]
If my project exists node_modules folder will appear cpu 100% of the situation

**Comment by lainekendall** [comment]
https://code.visualstudio.com/updates/v1_16
click update on left, the top of each page has download links for vs code

**Comment by newbienewbie** [comment]
Hi, the same bug occours on my Surface Pro3.

## Env
OS: win64 
VSCode Version: 1.17.0 
locale: Chinese

## How to reproduce

Openning a npm package containing a `node_modules` folder with VSCode 1.17.0 and doing nothing else ，the CPU  usage of `rg.exe` will quickly grows to 80% above .

If I choose to wait for a min , the `rg.exe` will exit silently and the usage of CPU will fall down to 3% .

**Comment by chrmarti** [comment]
Everyone seeing this: Is this right after starting the application or only after using Ctrl + P (Cmd + P on Mac) to open a file?

Could you also check what command parameters `rg` is running with?

**Comment by newbienewbie** [comment]
@chrmarti  Hi . sorry for the late reply.

>  Is this right after starting the application or only after using Ctrl + P (Cmd + P on Mac) to open a file?

It's right that the CPU usage grows quickly to 100%  after starting the vscode ,  but I can't reproduce the error when openning a file using Ctrl+P.

> what command parameters rg is running with

```cmd
"c:\Program Files\Microsoft VS Code\resources\app\node_modules\vscode-ripgrep\bin\rg" --files --hidden --case-sensitive -g *.csproj -g *.sln -g *.csx -g **/*.csproj -g **/*.sln -g **/*.csx --no-ignore --follow -- .
```

the Surface Pro3  OS: Microsoft Windows [版本 10.0.15063] x64

however , when I used my desktop in my office in the morning today,I could not reproduce the same bug on my win7 x64 desktop : win7 professional x64 SP1（ 6.1.7601 ）

After going home , I come back to my Surface Pro3 and try to disable extenstions to see whether it is caused by some extension. Here's my result:

with [ms-vscode.csharp](https://marketplace.visualstudio.com/items?itemName=ms-vscode.csharp) installed , the `rg.exe` will takes a huge amount of CPU usage.
however , if I disable the ms-vscode.csharp extention ,opening a npm package containing node_modules folder will not cause the bug.

@617450941 @cipchk  do you get the ms-vscode.csharp installed?

**Comment by 617450941** [comment]
My cpu no matter how long it is maintained at 100% in exchange for the old version did not appear this situation

**Comment by cipchk** [comment]
@newbienewbie 
thx, can be confirmed ms-vscode.csharp extension that will cause `rg.exe` of CPU usage, fixed this to disable it.

but not all projects, have to meet the following qualifications:

+ large composite project (includes FE & C#)
+ has `node_modules`

**Comment by chrmarti** [comment]
Could you run the following command in Powershell from your workspace folder and let me know what the output is?
```
(Measure-Command { & 'C:\Program Files\Microsoft VS Code\resources\app\node_modules\vscode-ripgrep\bin\rg.exe' --files --no-ignore --follow --hidden -- . | Measure-Object –Line | Out-Default }).TotalSeconds
```

**Comment by chrmarti** [comment]
Also, what is the output of the following when run in Powershell while the `rg.exe` are running?
```
(Get-WmiObject Win32_Process -Filter "name = 'rg.exe'").CommandLine
```

**Comment by newbienewbie** [comment]
@cipchk I don't have my surface pro3 at hand , so I can't test the command now . 

While on the desktop in my office , the first command will issue a lot of errors (os error 3) that complain about “File system loop found” and  “path not founding ”. 

To get a complete output  , you can redirect the stderr to a file like this : 
```
(Measure-Command { & 'C:\Program Files\Microsoft VS Code\resources\app\node_modules\vscode-ripgrep\bin\rg.exe' --files --no-ignore --follow --hidden -- . | Measure-Object –Line | Out-Default }).TotalSeconds 2>out.txt
```

can you load the complete out.txt ?

for the second command :

>> (Get-WmiObject Win32_Process -Filter "name = 'rg.exe'").CommandLine

the output is :
```cmd
"C:\Program Files\Microsoft VS Code\resources\app\node_modules\vscode-ripgrep\bin\rg.exe" --files --no-ignore --follow -
```

Is that the same to your output ?

**Comment by cipchk** [comment]
@newbienewbie Yes.

my steps:

1. run `(Get-WmiObject Win32_Process -Filter "name = 'rg.exe'").CommandLine`
2. run `(Measure-Command { & 'C:\Program Files\Microsoft VS Code\resources\app\node_modules\vscode-ripgrep\bin\rg.exe' --files --no-ignore --follow --hidden -- . | Measure-Object –Line | Out-Default }).TotalSeconds 2>out.txt`

> I confirmed that the run second command will 100% cpu!

so, the [out.txt](https://gist.github.com/cipchk/aa5a0bd5b6ed1af94b772226bdd57421) content.

**Comment by chrmarti** [comment]
Could you rerun the `Measure-Command...` command without `--follow`, that will help us understand if performance would be better when not following symlinks?

Please run the `Get-WmiObject` command while VSCode is showing the problem (not when you run the other command, I was imprecise in my instructions).

**Comment by cipchk** [comment]
@chrmarti There will be no `re.exe` 100% CPU usage if without `--follow`.

the all output content:

```
PS D:\psdm-FE> (Measure-Command { & 'C:\Program Files\Microsoft VS Code\resources\app\node_modules\vscode-ripgrep\bin\r
g.exe' --files --no-ignore --hidden -- . | Measure-Object –Line | Out-Default }).TotalSeconds

Lines Words Characters Property
----- ----- ---------- --------
84062


1.217515
```

the `--follow` parameter into a key problem?

**Comment by chrmarti** [comment]
@cipchk Which version of npm did you use to install the node_modules? I'm trying to reproduce the problem.

I'm not sure yet if removing `--follow` will be the best solution in your case.

**Comment by cipchk** [comment]
my env:

+ npm: 5.3.0
+ node: v8.5.0
+ system: win10 64bit
+ about vscode:

```
Version 1.17.0
Commit be377c0faf7574a59f84940f593a6849f12e4de7
Date 2017-10-04T23:45:46.804Z
Shell 1.7.7
Renderer 58.0.3029.110
Node 7.9.0
Architecture x64
```

and, the `D:\psdm-FE\package.json` content:

```json
{
  "name": "ps-h5",
  "version": "1.0.0",
  "author": "@cipchk",
  "dependencies": {},
  "devDependencies": {
    "babel-preset-es2015": "6.24.1",
    "event-stream": "3.3.2",
    "gulp": "3.9.1",
    "gulp-angular-htmlify": "2.3.0",
    "gulp-babel": "6.1.2",
    "gulp-bom": "1.0.0",
    "gulp-buffer": "0.0.2",
    "gulp-changed": "3.1.0",
    "gulp-clean-css": "3.5.0",
    "gulp-concat": "2.6.0",
    "gulp-if": "2.0.2",
    "gulp-inject": "4.2.0",
    "gulp-livereload": "3.8.0",
    "gulp-minify-html": "1.0.4",
    "gulp-modify-file": "^1.0.0",
    "gulp-ng-annotate": "2.0.0",
    "gulp-ng-html2js": "0.2.0",
    "gulp-rename": "1.2.2",
    "gulp-sass": "3.1.0",
    "gulp-sequence": "^0.4.6",
    "gulp-sync": "0.1.4",
    "gulp-uglify": "3.0.0",
    "gulp-util": "^3.0.6",
    "gulp-wrap": "0.13.0",
    "lodash": "4.17.4",
    "map-stream": "0.0.7",
    "merge-stream": "1.0.1",
    "merge2": "1.0.2",
    "path": "0.12.7",
    "rev-hash": "^2.0.0",
    "run-sequence": "^2.2.0",
    "sort-keys": "^2.0.0",
    "through2": "2.0.0",
    "vinyl-file": "^3.0.0"
  }
}
```

All of the above.

**Comment by cipchk** [comment]
@chrmarti 

I found reason, GOD!

Use the `cnpm install` command to install the package will have `rg.exe` 100% CPU usage, but works with `npm install`.

I think it's a file name problem, eg: `gulp` package should correspond to `gulp` fold, but `cnpm` actually use shortcuts to `_gulp@3.9.1@gulp` fold.

I don't know the question of `cnpm` or `vscode`.

**Comment by chrmarti** [comment]
I can reproduce it with `cnpm`. Thanks!

**Comment by chrmarti** [comment]
Looks like you could use `cnpm install --by=npm` to avoid the symlinks and work around the problem if you have to use `cnpm`.

**Comment by zoeDylan** [comment]
@chrmarti @cipchk @fengmk2 @dead-horse

use `npm install`
![by npm](https://user-images.githubusercontent.com/7030260/31418464-b3eb1104-ae68-11e7-83ed-a09fa602853a.jpg)

---

use `cnpm install`

![by cnpm](https://user-images.githubusercontent.com/7030260/31418477-c1730d72-ae68-11e7-90b9-bdf36e840b64.png)

Why?

I watched their folder contents。

**npm**     

674  projects

![by npm all file](https://user-images.githubusercontent.com/7030260/31418630-bac597e6-ae69-11e7-8b21-9fa8ca15d161.png)

---

**cnpm**   

1488  projects

![by cnpm all file](https://user-images.githubusercontent.com/7030260/31418631-c010d4cc-ae69-11e7-8477-ef0916c0214b.png)

look thes：*cnpm*

![cnpm](https://user-images.githubusercontent.com/7030260/31418719-4dc01102-ae6a-11e7-9226-f04e58f3ada7.png)

It has a lot of shortcuts，Don't know whether I can optimize performance from here；

**Comment by dead-horse** [comment]
hi guys, cnpm use symbol link to speed up installation like [pnpm](https://github.com/pnpm/pnpm/blob/master/docs/pnpm-vs-npm.md), so can you found a way to fix this issue not by using `cnpm install --by=npm` ?

**Comment by zoeDylan** [comment]
@dead-horse   Thanks！

**Comment by chrmarti** [comment]
Tracking this on ripgrep's side with https://github.com/BurntSushi/ripgrep/issues/633.

**Comment by Pacu2** [comment]
Same for me on Ubuntu, 100% CPU usage spikes just from typing, file search is painfully slow

**Comment by chrmarti** [comment]
@Pacu2 Could you open a new issue using Help > Report Issues?

**Comment by gy18505** [comment]
so what is the the best way to solve this problem? just simply use cnpm install --by=npm?

**Comment by chrmarti** [comment]
@gy18505 Yes, for now `cnpm install --by=npm` seems to be the simplest workaround.

**Comment by rvong** [comment]
Same on MacOS using SSHFS/OSXFUSE mounted folder.

**Comment by chrmarti** [comment]
The upcoming 1.18 release has a new setting `"search.followSymlinks"` that you can set to false as an alternative workaround for this issue. (From #37000.)

**Comment by zhaocun6592** [comment]
下载 vscode1.18以上版本  Documentation for Visual Studio Code 1.18

安装完成后 文件>设置>首选项     搜索设置 search.followSymlinks ：false；

**Comment by Hideer** [comment]
将vscode 中的 search.followSymlinks:设置为false亲测可行

**Comment by mccreeker** [comment]
Same thing happened after re-cloning a repo folder and opening VS Code in it:

![image](https://user-images.githubusercontent.com/33667567/35456436-a6321a16-0292-11e8-84ac-8b0df8320ac4.png)

Windows 10 Task Manager says the file spiking the CPU for several minutes (and slowing down the computer) is C:\Program Files\Microsoft VS Code\resources\app\node_modules\vscode-ripgrep\bin\rg.exe

ripgrep should really not go above 50% CPU, it should limit itself.

**Comment by pgonzal** [comment]
I am having this same issue.  The command line looks like this:

> `"c:\Program Files\Microsoft VS Code\resources\app\node_modules\vscode-ripgrep\bin\rg" --files --hidden --case-sensitive -g /*.csproj -g /*.sln -g /*.csx -g /*.cake -g **/*.csproj -g **/*.sln -g **/*.csx -g **/*.cake --no-ignore --follow --quiet -- .`


This operation is completely inappropriate in my environment. By design, the `node_modules` subfolder of my project contains thousands of symlinks into thousands of nested folders.  It would never make sense for a tool to do a brute force search of this subfolder without an explicit request from the developer.

In my repro, I did not perform any searches using the search tab.  I simply launch VS Code and it immediately spawns rg.exe.

Has anyone figured out **why** VS Code is doing this?  It is it a plugin?  Is it a misguided feature of VS Code itself?

**Comment by pgonzal** [comment]
Workaround:  Deleting rg.exe seems to eliminate the problem without any apparent loss of functionality in VS Code.

**Comment by warpdesign** [comment]
I am running into the same probelm: the second I press ctrl+c it starts eating cpu.

What's cnpm? I just opened a directory using VSCode and have no idea what cnpm is.

**Comment by warpdesign** [comment]
@pgonzal Deleting rg.exe will break quick open feature.

**Comment by warpdesign** [comment]
Setting `search.followSymLinks` to `false` doesn't help.

**Comment by pcwikonmap** [comment]
I also have the issue of VScode running at 100% cpu when doing a file search.

**Comment by superouman** [comment]
I also have the issue of VScode running at 100% cpu when doing a file search.

**Comment by pgonzal** [comment]
@BurntSushi is this the same issue that you fixed in BurntSushi/ripgrep#633?  If so does VS Code just need to upgrade?

**Comment by BurntSushi** [comment]
@pgonzal I think unfortunately there isn't a simple answer to that question. I guess the closest answer is "wontfix" or more like "unable to fix." There is just a fundamental limit on how fast a tool can process a directory tree containing 2 million entries. You'll need to read [my comment](https://github.com/BurntSushi/ripgrep/issues/633#issuecomment-362134936) to understand this issue from ripgrep's end. (Others are reporting issues even when symlinks are disabled, but I need a reproducible case outside of VS code on my end.)

**Comment by warpdesign** [comment]
I think I found a workaround: replacing the `rg.exe` binary by version `0.6.0` inside VS Code resources directory seems to have fixed the problem.

**Comment by BurntSushi** [comment]
@warpdesign Now _that_ would be interesting. Can anyone provide the exact command being run and a corpus to reproduce it with?

**Comment by chrmarti** [comment]
What might explain the renewed interest here: We updated to ripgrep 0.7.1 for the VS Code 1.20 released last week. VS Code 1.19 was based on 0.6.0.

**Comment by chrmarti** [comment]
@pgonzal The command arguments you posted indicate that the file search was triggered by an extension for C# that would activate for certain file types. Could you disable that extension and see if that works around the problem?

**Comment by chrmarti** [comment]
@pgonzal One other thing to try first would be to make sure you have `search.followSymlinks` set to `false` in your settings. You command line arguments show `--follow` which is added to follow symlinks.

**Comment by chrmarti** [comment]
@warpdesign Could you try running the original `rg` with the same command line arguments you posted from your project folder? Does that stop or slow down at some point? Any errors being shown?

**Comment by roblourens** [comment]
This issue is for a specific issue affecting people using cnpm/pnpm. The fix for that will probably come with https://github.com/Microsoft/vscode/issues/34711 and with extensions not making wide-ranging searches over all of node_modules.

@warpdesign and anyone not using those tools, please open a new issue with specific repro steps. Also if you put old rg.exe into insiders 1.21, it will break, so make sure that search/quickopen works at all.

**Comment by shrivatsahosabettu** [comment]
rg.exe is using 100% disk space also. Please let me know if there is any alternative, or is this is going to fix soon?

![image](https://user-images.githubusercontent.com/12061573/36311316-e3d1f700-1350-11e8-87a4-9255623b420f.png)

**Comment by BurntSushi** [comment]
@shrivatsahosabettu AIUI, that's not disk space, but disk bandwidth. And rg.exe is only using a paltry `0.4MB/s` in your screenshot.

**Comment by chrmarti** [comment]
I think we have addressed this issue with QuickOpen and text search when there are a lot of symlinks (cnpm/pnpm) in the folder being searched by adding the setting `search.followSymlinks`. The `workspaceContains` trigger will be covered with #34711.

For any remaining issues, especially regressions between releases, please open a new issue so we can follow up with you on the details. Thanks.

**Comment by warpdesign** [comment]
@chrmarti I think you are closing the issue too quickly: as I said above, adding `search.followSymlinks: false` doesn't fix the bug for me. And I don't have that much symlinks.

Also, I **need** to be able to search through symlinks.

**Comment by shrivatsahosabettu** [comment]
@burntsushi. In the screen print only 2 items are running as of now. But some more are running below that also. One more thing I noticed is this will run for first 5-6 min of my VS Code launch where the disk will become 100%, after 5-6 min all the rg.exe will disappear and disk size will come down to 36-40.

**Comment by chrmarti** [comment]
@warpdesign Let's continue the investigation in a new GitHub issue. It sounds like your problem is not the same we discussed and fixed here.

When you open a new issue could also do the following and include the result: Try running the original `rg` with the same command line arguments you posted above in your project folder? Does that stop or slow down at some point? Any errors being shown?

Thanks.

**Comment by chrmarti** [comment]
@shrivatsahosabettu If this reproduces also with the setting `"search.followSymlinks": false`, please open a new issue so we can follow up with you on the details. Thanks.
